src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/VisualStudioWindowsProjectCapabilityResolver.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Editor.Razor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    [Export(typeof(ProjectCapabilityResolver))]
+    internal class VisualStudioWindowsProjectCapabilityResolver : ProjectCapabilityResolver
+    {
+        public override bool HasCapability(string documentFilePath, object project, string capability)
+        {
+            if (project is not IVsHierarchy vsHierarchy)
+            {
+                return false;
+            }
+
+            var localHasCapability = LocalHasCapability(vsHierarchy, capability);
+            return localHasCapability;
+        }
+
+        private static bool LocalHasCapability(IVsHierarchy hierarchy, string capability)
+        {
+            if (hierarchy is null)
+            {
+                return false;
+            }
+
+            try
+            {
+                var hasCapability = hierarchy.IsCapabilityMatch(capability);
+                return hasCapability;
+            }
+            catch (NotSupportedException)
+            {
+                // IsCapabilityMatch throws a NotSupportedException if it can't create a
+                // BooleanSymbolExpressionEvaluator COM object
+                return false;
+            }
+            catch (ObjectDisposedException)
+            {
+                // IsCapabilityMatch throws an ObjectDisposedException if the underlying hierarchy has been disposed
```

Comment:
I feel like being handed objects which are (or were soon to be when handed off) disposed is worthy of a log.

Reply:
I thinkkkkk this was for teardown scenarios. I can see if I can import the Razor logger here to do the needful